Add PLAID optimizations to ColBERT

santhnm2 opened the pull request.

franzmpub commented on primeqa/ir/dense/colbert_top/colbert/indexing/codecs/residual.py:
```diff
@@ -166,51 +276,24 @@ def decompress(self, compressed_embs: Embeddings):
             We batch below even if the target device is CUDA to avoid large temporary buffers causing OOM.
         """
 
+        assert self.use_gpu
```
OK, but there are many instances of `if self.use_gpu` above.

ma-sultan commented on primeqa/ir/dense/colbert_top/colbert/modeling/colbert.py:
```diff
@@ -130,7 +167,9 @@ def doc(self, input_ids, attention_mask, keep_dims=True):
-        D = torch.nn.functional.normalize(D, p=2, dim=2).half()
+        D = torch.nn.functional.normalize(D, p=2, dim=2)
+        if self.use_gpu:
+            D = D.half()
```
D = D.cuda()?

franzmpub replied:
I think it's on a GPU already if a GPU is used.

franzmpub commented on primeqa/ir/dense/colbert_top/colbert/modeling/colbert.py:
```diff
@@ -28,6 +36,35 @@ def __init__(self, name='bert-base-uncased', colbert_config=None):
+                verbose=verbose,
+            )
+        except (RuntimeError, KeyboardInterrupt) as e:
+            if not verbose:
```
Printing stack if_not_ verbose?

santhnm2 replied:
If we set the verbose flag on the `load` call then I believe it prints the stack regardless of the try/catch

franzmpub replied:
OK, makes sense.

franzmpub commented on primeqa/ir/dense/colbert_top/colbert/search/index_loader.py:
```diff
@@ -19,17 +21,20 @@ def __init__(self, index_path):
         else:
-            # ivf, ivf_lengths = ivf.cuda(), torch.LongTensor(ivf_lengths).cuda()  # FIXME: REMOVE THIS LINE!
-            ivf = StridedTensor(ivf, ivf_lengths)
+            assert os.path.exists(os.path.join(self.index_path, "ivf.pt"))
```
It might be good to have messages in more of the asserts, e.g.
> assert os.path.exists(os.path.join(self.index_path, "ivf.pt")), f"ivf.pt not found in {self.index_path}"

santhnm2 replied:
Good point, will add these

franzmpub commented on primeqa/ir/dense/colbert_top/colbert/search/index_storage.py:
```diff
@@ -9,38 +10,72 @@
+                verbose=verbose,
+            )
+        except (RuntimeError, KeyboardInterrupt) as e:
+            if not verbose:
```
Printing stack if not verbose?
But, it's the same logic as in 

> modeling/colbert.py:59:            if not verbose:
search/index_storage.py:50:            if not verbose:
search/index_storage.py:70:            if not verbose:
indexing/codecs/residual.py:129:            if not verbose:
indexing/codecs/residual.py:152:            if not verbose:

franzmpub commented on setup.py:
```diff
@@ -31,7 +31,6 @@
     "bitarray~=2.3.7": ["install"],
     "bump2version~=1.0.1": ["dev"],
     "click~=8.0.4": ["install"],
-    "cupy-cuda113~=10.4.0": ["install"],
```
I'm not sure what removing 
`cupy-cuda113~=10.4.0": ["install"],`
will do to the rest of the components in PrimeQA, but the CI tests are passing.

franzmpub reviewed (commented):
LGTM, with emphasis on the "L", given the volume of code and functionalities added/changed.
The reassuring part is that 
1. Most of the code has been verified by multiple experiments before moving it to PrimeQA.
2. I was able to get results very close to the baseline ColBERT V2 search running on XOR-TyDi.

bhavani105 reviewed (commented):
Some initial thought:
- Provide some documentation on the arguments/paramters particularly highlighting the changes from ColBERT v2
- Test the ir/dense notebooks

avisil reviewed (changes requested):
LGTM @santhnm2 . Before merging I'd like to see:
1. I added one comment for adding documentation to the params for the end-user as it's not trivial in the current version. 
2. Also, please work with @franzmpub to make sure the jupyter notebook for ColBERT is working .

santhnm2 commented:
> LGTM @santhnm2 . Before merging I'd like to see:
> 
> 1. I added one comment for adding documentation to the params for the end-user as it's not trivial in the current version.
> 2. Also, please work with @franzmpub to make sure the jupyter notebook for ColBERT is working .

I updated `primeqa/ir/README.md` with a section on the PLAID hyperparameters.

franzmpub commented:
>     2. Also, please work with @franzmpub to make sure the jupyter notebook for ColBERT is working .

I've updated the jupyter notebook for ColBERT on 8/11/22

franzmpub commented:
@avisil I've checked the description of the PLAID parameters in the README - it's quite brief, but it servers the purpose.

bhavani105 reviewed (approved):
LGTM - Based on my limited usage of the indexing and search functions

One request:
The cast of qid to int does not seem necessary here - can we take the line out:
https://github.com/primeqa/primeqa/blob/cecbd5259eba4ccfdf00fdd961199f28135d2a3c/primeqa/ir/dense/colbert_top/colbert/evaluation/loaders.py#L23

avisil reviewed (approved):
It seems the current level of documentation will work for the merge.
